sklearn/neighbors/nth_element.pyx

```diff
@@ -0,0 +1,9 @@
+from typedefs cimport DTYPE_t, ITYPE_t
+
+cdef extern from "nth_element.h":
```

Comment:
Why do we not just define this extern directly in `binary_tree.pxi`?

Reply:
My original definition is fine to be put in `binary_tree.pxi`. But in my recent changes `partition_node_indices` is based on template now. The new `nth_element.pyx` is actually a bridge between Cython and C++ code. Without this, I'll have to change `_kd_tree.pyx` and `_ball_tree.pyx` to C++ mode.

Reply:
> Without this, I'll have to change _kd_tree.pyx and _ball_tree.pyx to C++ mode.

Excuse my ignorance... What risks come with changing them to C++ mode?